src/os/README.md

```diff
@@ -104,3 +106,86 @@ inspiration for several decisions in this library.
 - that said, if an operation is extremely efficient and almost always wanted,
   consider doing it natively, automatically. For example, ints in HSL
   `sockaddr` are always in host byte order, not network byte order.
+
+## Hack representation of C data types 
+### Cheat sheet
+
+| C type | Hack type |
+|---|---|
+| Pointers to referential transparent C `struct`s | `vec` |
+| Other C handles or pointers | `HH\Lib\OS\FileDescriptor` or other wrapper classes |
+
+### Default encoding for handles or pointers
+
+The underlying OS API commonly exposes some data type as handlers or pointers,
+e.g. `int filedes`, along with functions to create, manipulate, and destroy
+them. By default, handlers and pointers should be wrapped in HSL as closeable
+Hack classes.
```

Comment:
remove 'closeable'

When resource freeing is observable:
- should be closeable
- destructor must not free the resource
- closing should free the resource
- if not closed, should be freed at end of the request, or by a garbage collector

If resource freeing is not observable, we should not expose it at all; the intent here is to give us as much flexibility as possible for alternative garbage collection approaches in the future.

Reply:
The resource freeing is always observable for a pointer or handle. After all, we have to release the memory.

Reply:
The memory management of hack programs is considered unobservable - there is no way - beyond debug APIs - to be sure that memory allocation changes correspond to the hack code.

HHVM is free to optimize any mallocs/frees to different points in time; File handles are different as it whether or not a file is closed leads to clear different behavior in later function calls.

Reply:
>  After all, we have to release the memory even when there is no other side effects.

This, in particular, is not guaranteed for any hack code, especially not in timing

Reply:
I am not sure if we are on the same page. My understanding is that any C pointer like `xxx *` held by Hack would require the Hack user to explicitly call `close` or `using` to get released. This is the default case described in this section. Therefore, in this section I described them as "closeable".

On the other hand, the `vec` or `shape` encoding can be considered as a special optimization to avoid Hack program explicitly holding the C pointer, described in other sections.

Reply:
No - that just makes them be released *early*; if something is not closed, HHVM will clean it up 'some time'; when this happens is - and needs to remain - an implementation detail.

For most things, it's currently as soon as the refcount is 0 (and, for this reason, HHVM internally has a reference count for all objects)

We removed `__destruct()` from hack as we intend for this reference counting to be an 'unobservable' implementation detail and people were often depending on things that were happening in the destructors.

This also goes for `resource` types, like the ones you get from calling the PHP resources. This turned out to be a blocker on further experimentation with the garbage collection: people would open a file, never explicitly close it, and depend on it having been implicitly closed (which in turn implies a flush) before they re-opened it.

For `OS\FileDescriptor`, it's currently at the end of the request to make it so that people *can not* write code that depends on the implicit close.

It's best practice to use `using`, but if you never call `close()` or `closeWhenDisposed()`, that's fine too

Reply:
We also still have `__destruct`-on-refcount=0 in hack for objects that are implemented in C++; we can use it, but we don't want it to contain behavior that is 'useful' to hack programmers, as that will lead to programs depending more heavily on reference counting

Reply:
I see. So if the side effect is just memory release, we can create a C++ wrapped object with a hidden `__destruct`, right?